ocaml/xapi/repository_helpers.ml

```diff
@@ -1478,3 +1524,114 @@ let set_pending_guidances ~ops ~coming =
            to_be_removed ;
          do_with_vm_pending_guidances ~op:ops.vm_add ~vm:vm_ref_str to_be_added
      )
+
```

Comment:
I think the above function `set_pending_guidances` is a little confusing:
`
let set_pending_guidances ~ops ~coming =
  let pending_of_host =
    ops.host_get () |> List.filter (fun x -> not (is_livepatch_failure x))
  in
  let to_be_added, to_be_removed =
    merge_with_pending_guidances ~pending:pending_of_host ~coming
  in
  do_with_host_pending_guidances ~op:ops.host_remove to_be_removed ;
  do_with_host_pending_guidances ~op:ops.host_add to_be_added ;
  ops.vms_get ()
  |> List.map (fun (vm_ref_str, pending_of_vm) ->
         let pending = List.append pending_of_host pending_of_vm in
         (vm_ref_str, merge_with_pending_guidances ~pending ~coming)
     )
  |> List.iter (fun (vm_ref_str, (to_be_added, to_be_removed)) ->
         do_with_vm_pending_guidances ~op:ops.vm_remove ~vm:vm_ref_str
           to_be_removed ;
         do_with_vm_pending_guidances ~op:ops.vm_add ~vm:vm_ref_str to_be_added
     )
`

The `to_be_added` for host guidance also contains vm guidance, and `to_be_added` for vm guidance also contains host guidance, the correctness now is implicitly guaranteed by `do_with_host_pending_guidances` and `do_with_vm_pending_guidances`.

I think calculate `to_be_added` as host guidance only before `do_with_host_pending_guidances` and calculate `to_be_added` as vm guidance only before `do_with_vm_pending_guidances` would make the code easier to understand and maintain.

Reply:
A VM guidance tasks may be impacted by a host guidance task. So before setting VM guidance, both VM and host guidance are calculated together. And eventually the  `type pending_ops` can ensure the guidance can be distinguished correctly between VM and host.